pkg/networking/middleware/response.go

```diff
@@ -0,0 +1,123 @@
+package middleware
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/snyk/error-catalog-golang-public/snyk"
+	"github.com/snyk/error-catalog-golang-public/snyk_errors"
+)
+
+type ResponseMiddleware struct {
+	next       http.RoundTripper
+	errHandler func(error, context.Context) error
+}
+
+func NewReponseMiddleware(roundTriper http.RoundTripper, errHandler func(error, context.Context) error) *ResponseMiddleware {
+	return &ResponseMiddleware{
+		next:       roundTriper,
+		errHandler: errHandler,
+	}
+}
+
+func errFromResponse(res *http.Response) error {
+	if res.StatusCode >= 200 && res.StatusCode < 300 {
+		return nil
+	}
+
+	if err := errFromStatusCode(res.StatusCode); err != nil {
+		return err
+	}
+
+	if err := errFromJSONAPI(res); err != nil {
+		return err
+	}
+
+	return snyk_errors.Error{
+		Title:       "Unsuccessful network request",
+		Description: "A network request failed because with the resulting status code.",
+		StatusCode:  res.StatusCode,
+		Detail:      "Use the `-d` flag to inspect the request and response.",
+		Level:       "error",
+	}
+}
+
+func errFromStatusCode(code int) error {
+	switch code {
+	case http.StatusUnauthorized:
+		return snyk.NewUnauthorisedError("Use `snyk auth` to authenticate.")
+	case http.StatusBadRequest:
+		return snyk.NewBadRequestError("The request cannot be processed.")
+	case http.StatusInternalServerError:
+		return snyk.NewServerError("Internal server error.")
+	default:
+		return nil
+	}
+}
+
+func errFromJSONAPI(res *http.Response) error {
+	snykErr := &jsonAPIErroResponse{}
+	defer res.Body.Close()
+
+	err := json.NewDecoder(res.Body).Decode(snykErr)
+	if err != nil {
+		return nil
+	}
+
+	return snykErr.ToErrorCatalog()
+}
+
+func (rm ResponseMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	res, err := rm.next.RoundTrip(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = errFromResponse(res)
+
+	if rm.errHandler != nil {
+		err = rm.errHandler(err, res.Request.Context())
+	}
+
+	// RoundTrip should return one or the other.
+	if err != nil {
+		res = nil
+	}
+
+	return res, err
+}
+
+type jsonAPIErroResponse struct {
```

Comment:
Suggestion: The logic and the data structures to decode error catalog errors should be living in error catalog.

Reply:
I looked and found out the logic for this is [included](https://github.com/snyk/error-catalog/blob/2d9a7fd6f71d7eddb940a4d00e6bd9269348b864/packages/error-catalog-golang/src/snyk_errors/json_api_error.go#L20) in the Error Catalog, but for some reason it's private. I could raise a PR to make this public.